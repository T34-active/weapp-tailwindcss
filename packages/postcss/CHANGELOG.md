# @weapp-tailwindcss/postcss

## 1.0.10-alpha.1

### Patch Changes

- [`23c08c7`](https://github.com/sonofmagic/weapp-tailwindcss/commit/23c08c7d7ad276035ffa3ef78d0fb00e1c717ddf) Thanks [@sonofmagic](https://github.com/sonofmagic)! - fix: isTailwindcssV4ModernCheck

## 1.0.10-alpha.0

### Patch Changes

- [`0d6564a`](https://github.com/sonofmagic/weapp-tailwindcss/commit/0d6564aaaea588289dac5b1b784b1902105b297a) Thanks [@sonofmagic](https://github.com/sonofmagic)! - fix: 优化 isTailwindcssV4ModernCheck 判断

## 1.0.9

### Patch Changes

- [`3113053`](https://github.com/sonofmagic/weapp-tailwindcss/commit/31130538a13098e2a1d29bbd331dc67a195689cf) Thanks [@sonofmagic](https://github.com/sonofmagic)! - Fix: Support `Tailwindcss@4.1.1` and fix [#619](https://github.com/sonofmagic/weapp-tailwindcss/issues/619)

## 1.0.8

### Patch Changes

- [`e8c4534`](https://github.com/sonofmagic/weapp-tailwindcss/commit/e8c4534ab588eae1115dac15b529b7a122141316) Thanks [@sonofmagic](https://github.com/sonofmagic)! - feat: 添加 `cssRemoveProperty` 选项，默认值为 `true`，这是为了在 `tailwindcss` 中移除这种 css 节点:

  ```css
  @property --tw-content {
    syntax: "*";
    initial-value: "";
    inherits: false;
  }
  ```

  这种样式在小程序中，没有任何的意义。

## 1.0.7

### Patch Changes

- [`e0c37f5`](https://github.com/sonofmagic/weapp-tailwindcss/commit/e0c37f5f546b143341a75701a1907f876df38fa9) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore(deps): upgrade

- Updated dependencies [[`e0c37f5`](https://github.com/sonofmagic/weapp-tailwindcss/commit/e0c37f5f546b143341a75701a1907f876df38fa9)]:
  - @weapp-tailwindcss/shared@1.0.2

## 1.0.7-alpha.0

### Patch Changes

- [`e0c37f5`](https://github.com/sonofmagic/weapp-tailwindcss/commit/e0c37f5f546b143341a75701a1907f876df38fa9) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore(deps): upgrade

- Updated dependencies [[`e0c37f5`](https://github.com/sonofmagic/weapp-tailwindcss/commit/e0c37f5f546b143341a75701a1907f876df38fa9)]:
  - @weapp-tailwindcss/shared@1.0.2-alpha.0

## 1.0.6

### Patch Changes

- [`ff9933a`](https://github.com/sonofmagic/weapp-tailwindcss/commit/ff9933ad06de7bf3333c1c63016920639a56b87a) Thanks [@sonofmagic](https://github.com/sonofmagic)! - fix: tailwindcss@4 space-x-_ and space-y-_

## 1.0.5

### Patch Changes

- [`a4532ab`](https://github.com/sonofmagic/weapp-tailwindcss/commit/a4532ab34de62556e57ed350e15ca14e602b7f93) Thanks [@sonofmagic](https://github.com/sonofmagic)! - fix: 使用space-y-2后编译报错 #595

## 1.0.4

### Patch Changes

- [`9e65534`](https://github.com/sonofmagic/weapp-tailwindcss/commit/9e65534f035ee4e17a2dc0b891278cacb92d5a0b) Thanks [@sonofmagic](https://github.com/sonofmagic)! - fix: @layer as :not(#n) and @weapp-tailwindcss/merge switcher

## 1.0.3

### Patch Changes

- Updated dependencies [[`bdbca26`](https://github.com/sonofmagic/weapp-tailwindcss/commit/bdbca268738ef033a5789e8a6713608c4d599b61)]:
  - @weapp-tailwindcss/shared@1.0.1

## 1.0.3-alpha.0

### Patch Changes

- Updated dependencies [[`bdbca26`](https://github.com/sonofmagic/weapp-tailwindcss/commit/bdbca268738ef033a5789e8a6713608c4d599b61)]:
  - @weapp-tailwindcss/shared@1.0.1-alpha.0

## 1.0.2

### Patch Changes

- [`64c0189`](https://github.com/sonofmagic/weapp-tailwindcss/commit/64c018935732481ebe2f366e4136b4d3574dde57) Thanks [@sonofmagic](https://github.com/sonofmagic)! - feat: improve `isAllowedClassName` preflight

## 1.0.1

### Patch Changes

- [`ee34fb3`](https://github.com/sonofmagic/weapp-tailwindcss/commit/ee34fb34688a2bd11018ce5e4ea6d07a062b0b55) Thanks [@sonofmagic](https://github.com/sonofmagic)! - fix: bg-gradient --tw-gradient-position in oklab

## 1.0.0

### Major Changes

- [`bafd149`](https://github.com/sonofmagic/weapp-tailwindcss/commit/bafd149f0510b30cbf95711223583055023e7875) Thanks [@sonofmagic](https://github.com/sonofmagic)! - ## Feature

  增加 `@weapp-tailwindcss/merge` 支持，这是 `weapp-tailwindcss` 版本的 `tailwindcss-merge` 和 `cva` 方法

  ## Breaking Changes

  1. 去除 `weapp-tailwindcss/postcss` (可直接安装使用 `@weapp-tailwindcss/postcss`)
  2. 增加 `weapp-tailwindcss/escape` 来取代 `weapp-tailwindcss/replace`
  3. 项目 monorepo 区分包

### Patch Changes

- [`d4bfcc7`](https://github.com/sonofmagic/weapp-tailwindcss/commit/d4bfcc7a25776eb9869bd934c05e3b49a3f4cc8b) Thanks [@sonofmagic](https://github.com/sonofmagic)! - fix: remove attribute in :where pseudo

- [`2c51531`](https://github.com/sonofmagic/weapp-tailwindcss/commit/2c515310f1fdfd15d11e2e35213c7e6bfcb55c3d) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore(deps): upgrade

- [`d7fa028`](https://github.com/sonofmagic/weapp-tailwindcss/commit/d7fa02877ce74792687765766ff94ae3e30edf3b) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: rename export function

- [`1d689ef`](https://github.com/sonofmagic/weapp-tailwindcss/commit/1d689efca6cf0de7e476b03b2be8d09284beae68) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: change postcssPresetEnv default value

- [`1df699b`](https://github.com/sonofmagic/weapp-tailwindcss/commit/1df699bf82d66847cbc64cc8f294ef237e0470c5) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: fix types import intel

- [`6d20f1f`](https://github.com/sonofmagic/weapp-tailwindcss/commit/6d20f1f9a799cf350dcbbd861907f0ff70a68dfd) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: perfer postcss-preset-env instead of @csstools

- [`e745ba5`](https://github.com/sonofmagic/weapp-tailwindcss/commit/e745ba5cd9e232c0b5b7053538beb0772240eab8) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: bump version

- Updated dependencies [[`bafd149`](https://github.com/sonofmagic/weapp-tailwindcss/commit/bafd149f0510b30cbf95711223583055023e7875), [`2c51531`](https://github.com/sonofmagic/weapp-tailwindcss/commit/2c515310f1fdfd15d11e2e35213c7e6bfcb55c3d), [`d7fa028`](https://github.com/sonofmagic/weapp-tailwindcss/commit/d7fa02877ce74792687765766ff94ae3e30edf3b), [`d50d95d`](https://github.com/sonofmagic/weapp-tailwindcss/commit/d50d95d04e1c6b7c3ff32acc0d9894d3c0f06d22), [`e745ba5`](https://github.com/sonofmagic/weapp-tailwindcss/commit/e745ba5cd9e232c0b5b7053538beb0772240eab8)]:
  - @weapp-tailwindcss/shared@1.0.0

## 1.0.0-alpha.8

### Patch Changes

- [`1d689ef`](https://github.com/sonofmagic/weapp-tailwindcss/commit/1d689efca6cf0de7e476b03b2be8d09284beae68) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: change postcssPresetEnv default value

## 1.0.0-alpha.7

### Patch Changes

- [`d4bfcc7`](https://github.com/sonofmagic/weapp-tailwindcss/commit/d4bfcc7a25776eb9869bd934c05e3b49a3f4cc8b) Thanks [@sonofmagic](https://github.com/sonofmagic)! - fix: remove attribute in :where pseudo

## 1.0.0-alpha.6

### Patch Changes

- [`e745ba5`](https://github.com/sonofmagic/weapp-tailwindcss/commit/e745ba5cd9e232c0b5b7053538beb0772240eab8) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: bump version

- Updated dependencies [[`e745ba5`](https://github.com/sonofmagic/weapp-tailwindcss/commit/e745ba5cd9e232c0b5b7053538beb0772240eab8)]:
  - @weapp-tailwindcss/shared@1.0.0-alpha.4

## 1.0.0-alpha.5

### Patch Changes

- [`6d20f1f`](https://github.com/sonofmagic/weapp-tailwindcss/commit/6d20f1f9a799cf350dcbbd861907f0ff70a68dfd) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: perfer postcss-preset-env instead of @csstools

## 1.0.0-alpha.4

### Patch Changes

- [`d7fa028`](https://github.com/sonofmagic/weapp-tailwindcss/commit/d7fa02877ce74792687765766ff94ae3e30edf3b) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: rename export function

- Updated dependencies [[`d7fa028`](https://github.com/sonofmagic/weapp-tailwindcss/commit/d7fa02877ce74792687765766ff94ae3e30edf3b)]:
  - @weapp-tailwindcss/shared@1.0.0-alpha.3

## 1.0.0-alpha.3

### Patch Changes

- [`2c51531`](https://github.com/sonofmagic/weapp-tailwindcss/commit/2c515310f1fdfd15d11e2e35213c7e6bfcb55c3d) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore(deps): upgrade

- Updated dependencies [[`2c51531`](https://github.com/sonofmagic/weapp-tailwindcss/commit/2c515310f1fdfd15d11e2e35213c7e6bfcb55c3d)]:
  - @weapp-tailwindcss/shared@1.0.0-alpha.2

## 1.0.0-alpha.2

### Patch Changes

- [`1df699b`](https://github.com/sonofmagic/weapp-tailwindcss/commit/1df699bf82d66847cbc64cc8f294ef237e0470c5) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: fix types import intel

## 1.0.0-alpha.1

### Major Changes

- [`bafd149`](https://github.com/sonofmagic/weapp-tailwindcss/commit/bafd149f0510b30cbf95711223583055023e7875) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: monorepo changes

### Patch Changes

- Updated dependencies [[`bafd149`](https://github.com/sonofmagic/weapp-tailwindcss/commit/bafd149f0510b30cbf95711223583055023e7875)]:
  - @weapp-tailwindcss/shared@1.0.0-alpha.1

## 0.0.1-alpha.0

### Patch Changes

- Updated dependencies [[`d50d95d`](https://github.com/sonofmagic/weapp-tailwindcss/commit/d50d95d04e1c6b7c3ff32acc0d9894d3c0f06d22)]:
  - @weapp-tailwindcss/shared@0.0.1-alpha.0
