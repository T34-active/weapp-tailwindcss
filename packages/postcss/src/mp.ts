import type { IStyleHandlerOptions } from './types'
import { Declaration, Rule } from 'postcss'
import cssVarsV3 from './cssVarsV3'
import cssVarsV4 from './cssVarsV4'
import { isOnlyBeforeAndAfterPseudoElement } from './selectorParser'

const cssVarsV3Nodes = cssVarsV3.map((x) => {
  return new Declaration({
    prop: x.prop,
    value: x.value,
  })
})

export const cssVarsV4Nodes = cssVarsV4.map((x) => {
  return new Declaration({
    prop: x.prop,
    value: x.value,
  })
})
// ':not(template) ~ :not(template)'
// ':not(template)~:not(template)'
// const regexp1 = /:not\(template\)\s*~\s*:not\(template\)/g
// :not([hidden])~:not([hidden])
// :not([hidden]) ~ :not([hidden])
// const regexp2 = /:not\(\[hidden\]\)\s*~\s*:not\(\[hidden\]\)/g

/**
 * 带有子选择器的
 * https://tailwindcss.com/docs/space
 * .space-x-4>:not([hidden])~:not([hidden])
 *
 * https://tailwindcss.com/docs/divide-width
 * .divide-x>:not([hidden])~:not([hidden])
 *
 * https://tailwindcss.com/docs/divide-color
 * .divide-blue-200>:not([hidden])~:not([hidden])
 * :is(.dark .dark\:divide-slate-700)>:not([hidden])~:not([hidden])
 *
 * https://tailwindcss.com/docs/divide-style
 * .divide-dashed>:not([hidden])~:not([hidden])
 *
 * 其中小程序里直接写
 * .divide-y-4>:not(hidden) + :not(hidden)
 * 会语法错误
 */

export function testIfVariablesScope(node: Rule, count = 2): boolean {
  if (isOnlyBeforeAndAfterPseudoElement(node)) {
    const nodes = node.nodes
    let c = 0
    for (const tryTestDecl of nodes) {
      if (tryTestDecl && tryTestDecl.type === 'decl' && tryTestDecl.prop.startsWith('--tw-')) {
        c++
      }
      if (c >= count) {
        return true
      }
    }
    return false
  }
  return false
}

export function testIfTwBackdrop(node: Rule, count = 2) {
  if (node.type === 'rule' && node.selector === '::backdrop') {
    const nodes = node.nodes
    let c = 0
    for (const tryTestDecl of nodes) {
      if (tryTestDecl && tryTestDecl.type === 'decl' && tryTestDecl.prop.startsWith('--tw-')) {
        c++
      }
      if (c >= count) {
        return true
      }
    }
    return false
  }
  return false
}

export function testIfRootHostForV4(node: Rule) {
  return node.type === 'rule' && node.selector.includes(':root') && node.selector.includes(':host')
}

export function makePseudoVarRule() {
  const pseudoVarRule = new Rule({
    // selectors: ['::before', '::after'],
    selector: '::before,::after',
  })
  pseudoVarRule.append(
    new Declaration({
      prop: '--tw-content',
      value: '""',
    }),
  )
  return pseudoVarRule
}

export function remakeCssVarSelector(selectors: string[], options: IStyleHandlerOptions) {
  const { cssPreflightRange, cssSelectorReplacement } = options
  if (
    cssPreflightRange === 'all' // 默认对每个元素都生效
    && !selectors.includes(':not(not)')
  ) {
    selectors.push(':not(not)')
  }
  if (cssSelectorReplacement) {
    if (Array.isArray(cssSelectorReplacement.universal)) {
      if (
        !cssSelectorReplacement.universal.every((x) => {
          return selectors.includes(x)
        })
        && !selectors.includes('*')
      ) {
        selectors.unshift('*')
      }
    }
    else if (
      typeof cssSelectorReplacement.universal === 'string'
      && !selectors.includes(cssSelectorReplacement.universal)
      && !selectors.includes('*')
    ) {
      selectors.unshift('*')
    }
  }

  return selectors
}

export function commonChunkPreflight(node: Rule, options: IStyleHandlerOptions) {
  const { ctx, cssInjectPreflight, injectAdditionalCssVarScope } = options
  // css vars scope
  // node.selector = remakeCombinatorSelector(node.selector, options)

  // 变量注入和 preflight
  if (testIfVariablesScope(node)) {
    ctx?.markVariablesScope(node)
    node.selectors = remakeCssVarSelector(node.selectors, options)
    node.before(makePseudoVarRule())
    if (typeof cssInjectPreflight === 'function') {
      node.append(...cssInjectPreflight())
    }
  }
  const isTailwindcss4 = options.majorVersion === 4
  if (injectAdditionalCssVarScope && (isTailwindcss4 ? testIfRootHostForV4(node) : testIfTwBackdrop(node))) {
    const syntheticRule = new Rule({
      selectors: ['*', '::after', '::before'],
      nodes: isTailwindcss4 ? cssVarsV4Nodes : cssVarsV3Nodes,
    })
    syntheticRule.selectors = remakeCssVarSelector(syntheticRule.selectors, options)
    node.before(syntheticRule)
    node.before(makePseudoVarRule())
    if (typeof cssInjectPreflight === 'function') {
      syntheticRule.append(...cssInjectPreflight())
    }
  }
}
