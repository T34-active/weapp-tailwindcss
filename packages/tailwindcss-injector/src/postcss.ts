import type { PluginCreator } from 'postcss'
import type { Config } from 'tailwindcss'
import type { Options } from './types'
import postcss from 'postcss'
import set from 'set-value'
import tailwindcss from 'tailwindcss'
import { loadConfig } from 'tailwindcss-config'
import { getConfig } from './config'
import { postcssPlugin } from './constants'
import { regExpTest, removeFileExtension } from './utils'
import { getDepFiles } from './wxml'

// function isObject(obj: any): obj is object {
//   return typeof obj === 'object' && obj !== null
// }
export type {
  Options,
}

const creator: PluginCreator<Partial<Options>> = (options) => {
  const { config, filter, directiveParams, cwd, extensions, insertAfterAtRulesNames, insertAfterComments } = getConfig(options)

  const extensionsGlob = `{${extensions.join(',')}}`

  return {
    postcssPlugin,
    plugins: [
      {
        postcssPlugin: `${postcssPlugin}:post`,
        async Once(root, helpers) {
          if (filter(root.source?.input)) {
            const atruleLastIndex = root.nodes.findLastIndex(x => x.type === 'atrule' && insertAfterAtRulesNames.includes(x.name))
            const commentLastIndex = root.nodes.findLastIndex(x => x.type === 'comment' && regExpTest(insertAfterComments, x.text))
            const lastIndex = Math.max(atruleLastIndex, commentLastIndex)
            for (const params of directiveParams) {
              const node = root.nodes.find(x => x.type === 'atrule' && x.params === params)
              if (!node) {
                const atRule = helpers.atRule({ name: 'tailwind', params })
                if (lastIndex < 0) {
                  root.prepend(atRule)
                }
                else {
                  root.insertAfter(lastIndex, atRule)
                }
              }
            }

            const cfg = typeof config === 'function' ? config(root.source?.input) : config
            let tailwindcssConfig: Config | undefined
            if (typeof cfg === 'string') {
              const result = await loadConfig({
                cwd,
                config: cfg,
              })
              if (result) {
                tailwindcssConfig = result.config
              }
            }
            else {
              tailwindcssConfig = (cfg ?? {
                content: [],
              }) as Config
            }

            if (tailwindcssConfig && root.source?.input && root.source.input.file) {
              const basename = removeFileExtension(root.source.input.file)
              set(tailwindcssConfig, 'content', [
                `${basename}.${extensionsGlob}`,
              ])

              // 分析模板
              const deps = await getDepFiles(`${basename}.wxml`)
              for (const dep of deps) {
                if (Array.isArray(tailwindcssConfig.content)) {
                  tailwindcssConfig.content.push(`${removeFileExtension(dep)}.${extensionsGlob}`)
                }
              }

              await postcss([
                tailwindcss(tailwindcssConfig),
              ]).process(root, {
                from: root.source.input.file,
              }).async()
            }
          }
        },
      },
    ],
  }
}
creator.postcss = true

export default creator
