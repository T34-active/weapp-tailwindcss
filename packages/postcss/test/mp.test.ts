import postcss from 'postcss'
import { commonChunkPreflight } from '@/mp'

describe('mp', () => {
  it('commonChunkPreflight case 0', () => {
    const root = postcss.parse('.foo { color: red; }')
    commonChunkPreflight(root.first as postcss.Rule, {})
    expect(root.toString()).toBe('.foo { color: red; }')
  })

  it('commonChunkPreflight case 1', () => {
    const root = postcss.parse(':root,:host { color: red; }')
    commonChunkPreflight(root.first as postcss.Rule, {})
    expect(root.toString()).toBe(':root,:host { color: red; }')
  })

  it('commonChunkPreflight case 2', () => {
    const root = postcss.parse(':root,:host { color: red; }')
    commonChunkPreflight(root.first as postcss.Rule, { injectAdditionalCssVarScope: true })
    expect(root.toString()).toBe(':root,:host { color: red; }')
  })

  it('commonChunkPreflight case 3', () => {
    const root = postcss.parse(':root,:host { color: red; }')
    commonChunkPreflight(root.first as postcss.Rule, { injectAdditionalCssVarScope: true, majorVersion: 4 })
    expect(root.toString()).toMatchSnapshot()
  })
})
