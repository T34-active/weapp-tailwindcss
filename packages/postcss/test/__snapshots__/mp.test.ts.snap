// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mp > commonChunkPreflight case 3 1`] = `
"*, ::after, ::before { --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-translate-z: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-scale-z: 1; --tw-rotate-x: ; --tw-rotate-y: ; --tw-rotate-z: ; --tw-skew-x: ; --tw-skew-y: ; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-space-x-reverse: 0; --tw-space-y-reverse: 0; --tw-border-style: solid; --tw-divide-x-reverse: 0; --tw-divide-y-reverse: 0; --tw-gradient-position: ; --tw-gradient-from: #0000; --tw-gradient-via: #0000; --tw-gradient-to: #0000; --tw-gradient-stops: ; --tw-gradient-via-stops: ; --tw-gradient-from-position: 0%; --tw-gradient-via-position: 50%; --tw-gradient-to-position: 100%; --tw-mask-linear: linear-gradient(#fff, #fff); --tw-mask-radial: linear-gradient(#fff, #fff); --tw-mask-conic: linear-gradient(#fff, #fff); --tw-mask-left: linear-gradient(#fff, #fff); --tw-mask-right: linear-gradient(#fff, #fff); --tw-mask-bottom: linear-gradient(#fff, #fff); --tw-mask-top: linear-gradient(#fff, #fff); --tw-mask-linear-position: 0deg; --tw-mask-linear-from-position: 0%; --tw-mask-linear-to-position: 100%; --tw-mask-linear-from-color: black; --tw-mask-linear-to-color: transparent; --tw-mask-radial-from-position: 0%; --tw-mask-radial-to-position: 100%; --tw-mask-radial-from-color: black; --tw-mask-radial-to-color: transparent; --tw-mask-radial-shape: ellipse; --tw-mask-radial-size: farthest-corner; --tw-mask-radial-position: center; --tw-mask-conic-position: 0deg; --tw-mask-conic-from-position: 0%; --tw-mask-conic-to-position: 100%; --tw-mask-conic-from-color: black; --tw-mask-conic-to-color: transparent; --tw-font-weight: ; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-opacity: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-drop-shadow-color: ; --tw-drop-shadow-alpha: 100%; --tw-drop-shadow-size: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ; --tw-duration: ; --tw-ease: ; --tw-content: ""; --tw-contain-size: ; --tw-contain-layout: ; --tw-contain-paint: ; --tw-contain-style: ; --tw-leading: ; --tw-tracking: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-outline-style: solid; --tw-text-shadow-color: ; --tw-text-shadow-alpha: 100%; --tw-shadow: 0 0 #0000; --tw-shadow-color: ; --tw-shadow-alpha: 100%; --tw-inset-shadow: 0 0 #0000; --tw-inset-shadow-color: ; --tw-inset-shadow-alpha: 100%; --tw-ring-color: ; --tw-ring-shadow: 0 0 #0000; --tw-inset-ring-color: ; --tw-inset-ring-shadow: 0 0 #0000; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-offset-shadow: 0 0 #0000; --tw-mask-top-from-position: 0%; --tw-mask-top-to-position: 100%; --tw-mask-top-from-color: black; --tw-mask-top-to-color: transparent; --tw-mask-right-from-position: 0%; --tw-mask-right-to-position: 100%; --tw-mask-right-from-color: black; --tw-mask-right-to-color: transparent; --tw-mask-bottom-from-position: 0%; --tw-mask-bottom-to-position: 100%; --tw-mask-bottom-from-color: black; --tw-mask-bottom-to-color: transparent; --tw-mask-left-from-position: 0%; --tw-mask-left-to-position: 100%; --tw-mask-left-from-color: black; --tw-mask-left-to-color: transparent; }
::before,::after { --tw-content: ""; }
:root,:host { color: red; }"
`;
